dia-batching-server/src/api/polygon.rs

```diff
@@ -0,0 +1,276 @@
+			prices.push(quotation);
+		}
+
+		Ok(prices)
+	}
+
+	pub fn is_supported(asset: &AssetSpecifier) -> bool {
+		Self::extract_source_currency(asset).is_some()
+	}
+
+	/// Extract the source currency from the asset pair.
+	/// We assume that the symbol contained in the `AssetSpecifier` is of the form <from>-<to>.
+	fn extract_source_currency(asset: &AssetSpecifier) -> Option<String> {
+		let (blockchain, symbol) = (asset.blockchain.as_str(), asset.symbol.as_str());
+		if blockchain.to_uppercase() != "FIAT" {
+			return None;
+		}
+
+		// We assume to receive a symbol of form <from>-<to> and we want to extract the <from> part
+		let parts: Vec<_> = symbol.split('-').collect();
+		if parts.len() != 2 {
+			return None;
+		}
+		let from_currency = parts.get(0)?;
+		Some(from_currency.to_uppercase())
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct PolygonPrice {
+	pub converted: Decimal,
+	pub from: String,
+	#[serde(rename = "initialAmount")]
+	pub initial_amount: i64,
+	pub last: serde_json::Value,
+	pub request_id: String,
+	pub status: String,
+	pub symbol: String,
+	pub to: String,
+}
+
+/// Polygon network client
+pub struct PolygonClient {
+	host: String,
+	api_key: String,
+}
+
+impl PolygonClient {
+	pub fn new(host: String, api_key: String) -> Self {
+		PolygonClient { host, api_key }
+	}
+
+	async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, PolygonError> {
+		let mut headers = reqwest::header::HeaderMap::new();
+		headers.insert("accept", reqwest::header::HeaderValue::from_static("application/json"));
+
+		let bearer = "Bearer ".to_string() + self.api_key.as_str();
+		let mut api_key_header = reqwest::header::HeaderValue::from_str(bearer.as_str())
+			.expect("Could not create header value");
+		api_key_header.set_sensitive(true);
+		headers.insert("Authorization", api_key_header);
+
+		let client = reqwest::Client::builder()
+			.default_headers(headers)
+			.build()
+			.map_err(|e| PolygonError(e.to_string()))?;
+
+		let url = reqwest::Url::parse(
+			format!("{host}/{ep}", host = self.host.as_str(), ep = endpoint).as_str(),
+		)
+		.expect("Invalid URL");
+
+		let response = client.get(url).send().await;
+
+		match response {
+			Ok(response) => {
+				if !response.status().is_success() {
+					let result = response.text().await;
+					Err(PolygonError(result.unwrap()))
+				} else {
+					let result = response.json().await;
+					result.map_err(|e| {
+						PolygonError(
+							"Could not decode Polygon response: ".to_owned() + &e.to_string(),
+						)
+					})
+				}
+			},
+			Err(e) => Err(PolygonError(e.to_string())),
+		}
+	}
+
+	/// Get the current price of any fiat currency in USD
+	pub async fn price(&self, from_currency: String) -> Result<PolygonPrice, PolygonError> {
+		// Currencies have to be upper-case
+		let from_currency = from_currency.to_uppercase();
+		// We always query for USD.
+		let to_currency = "USD";
+		// We always query for full precision
+		let precision = "5";
```

Comment:
Seem like the precision can be at most 4 from Polygon API [docs](https://polygon.io/docs/forex/get_v1_last_quote_currencies__from___to) or am I looking at the wrong doc page? 
Also I cannot query the `conversion` endpoint from their doc page to see what happens when using `5` as the value for `precision` since they require at least a paid Starter subscription to do that.

Reply:
Yes, that's what they put in their docs but I noticed that I can actually get higher precision by defining a larger value. The highest precision I could get was `5` though. You can even put a precision of `10` or similar but the response will not have more than 5 decimals as far as I can tell. That's why I chose that value here.